### ISSUE TITLE
REST dispatcher assumes the request is to the API

### ISSUE THREAD

**Issue by rmccue** [johnbillion/query-monitor#183]
If you're running the API in an embedded fashion (that is, calling `rest_get_server()->dispatch()`) in your code, Query Monitor errors by trying to load classes in multiple times on the page.

`QM_Dispatcher_REST::filter_rest_post_dispatch()` should have a check for the `REST_REQUEST` constant.

**Comment by johnbillion** [comment]
Thanks, I'll get this fixed.

Is there a chance that a registered route's callback function could call `rest_get_server()->dispatch()` to perform an internal API call? If so, `REST_REQUEST` would still be defined in this situation.

**Comment by rmccue** [comment]
> Is there a chance that a registered route's callback function could call `rest_get_server()->dispatch()` to perform an internal API call?

Yup; the various creation functions `POST /wp/v2/posts` run this internally.

If you need stuff to run once, `rest_api_init` is safe to use, as it only runs once. You can still check `REST_REQUEST` there for API-only requests.

**Comment by johnbillion** [comment]
Ryan, are you able to provide some example code which triggers this error? Before I fix the issue I want to reproduce it.

The code below does not trigger the error because `WP_REST_Server::dispatch()` doesn't call the `rest_post_dispatch` filter.

```php
$server = rest_get_server();
$server->dispatch( new WP_REST_Request( 'GET', '/wp/v2/categories' ) );
$server->dispatch( new WP_REST_Request( 'GET', '/wp/v2/posts' ) );
```

**Comment by rmccue** [comment]
`rest_post_dispatch` is also called [in `embed_links`](https://github.com/WordPress/WordPress/blob/944a69df87a6a7f50f403707a4bc99aa24c9b6cf/wp-includes/rest-api/class-wp-rest-server.php#L581), which is also internally usable. Something like this fails:

```php
$server = rest_get_server();
$response = $server->dispatch( new WP_REST_Request( 'GET', '/wp/v2/posts' ) );
$data = $server->response_to_data( $response, true );
```

**Comment by johnbillion** [comment]
Thanks!

**Comment by johnbillion** [comment]
Fixed in f0ea330 and 08ab4b0.
